GSOC'25 - MDEV-36100 Automatic embedding generation and "expensive" functions

apostolis1 opened the pull request:
This DRAFT PR is part of the GSOC'25 project "[MDEV-36100](https://jira.mariadb.org/browse/MDEV-36100) Generate vector embeddings automatically on INSERT".
This is a DRAFT PR, meant to be used for mentor feedback and will be updated as the project progresses.

There are two main aspects on this PR:
1. Plugin to automatically generate vector emebeddings
2. Avoidance of recalculation of stored expensive virtual columns

<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-36100*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
4. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
5. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description

TODO: This will be updated 

## Release Notes

The plugin uses the following session variables:
1. `generate_embedding_openai_host`: The API endpoint the request will be sent to
2. `generate_embedding_openai_api_key`: The API key used for the request, provided via [HTTP Bearer authentication](https://platform.openai.com/docs/api-reference/authentication) 

and provides the following status variables to help system admins keep track of requests:
1. `generate_embedding_openai_successful_http_requests`
2. `generate_embedding_openai_total_http_requests`


## How can this PR be tested?

This PR creates a new testing suite, `gen_embedding`,which includes tests both for the plugin and the expensive functionality.
Expensive tests have the prefix `expensive_`, to distinguish them from the plugin tests.

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and coding style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## Future Work

In terms of the plugin, the status variables are not being reset by `FLUSH STATUS`, it remains to be investigated why that is and whether there is special support needed from the plugin.

As far as the stored expensive virtual columns are concerned:

- Currently, changes in the virtual columns data type in `ALTER` statements are not detected.
So, something like the following incorrectly does not recompute the virtual column:
```sql
create table t (
    doc text,
    embedding vector(1536) as (generate_embedding_openai(doc, "text-embedding-ada-002")) stored
) engine=MyISAM;
alter table t modify embedding vector(2000);
```

- We would also like to avoid recomputing stored expensive virtual columns in replication setups whenever possible.
Initially, we would want to do this for [Row-Based Logging](https://mariadb.com/docs/server/server-management/server-monitoring-logs/binary-log/binary-log-formats#row-based-logging) and then for [Mixed Logging](https://mariadb.com/docs/server/server-management/server-monitoring-logs/binary-log/binary-log-formats#mixed-logging).
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

apostolis1 committed:
Initial version of the plugin. Does not support API calls and plugin variables

apostolis1 committed:
Updates the Plugin author

apostolis1 committed:
Fixes wrong identation causing compilation failure

apostolis1 committed:
Adds session variables for host and api key

apostolis1 committed:
Adds curl support to the plugin. Currently no user input is taken into consideration. The curl post works with an internal testing mock server, not with the actual openai server, as no authorization is used.

apostolis1 committed:
Adds a minimal Flask server and an API response from OpenAI, useful for testing the plugin

apostolis1 commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,356 @@
+    {
+      // TODO: I am grabbing the session variables here, because this is the function that has access to the THD
+      // Is there a better place to do this?
+      host = THDVAR(thd, host);
```
Regarding the plugin variables and access to the thread descriptor, I just came across `_current_thd`
This allows me to create a getter, for example
`const char* get_openai_host(THD* thd)
{
  return THDVAR(thd, host);
}`
which can then be used as `const char* myHost = get_openai_host(_current_thd());` where appropriate. 
I will use this pattern, which seems to be the correct way to handle plugin variables

apostolis1 replied:
For transparency: This has been discussed internally, and we agreed that it makes sense for `host` to be a session var

apostolis1 committed:
Implements parsing the function's input for the API call

apostolis1 committed:
Adds error handling for curl and http errors

apostolis1 committed:
Adds a map with supported models. This can be used to check whether the model name is valid and determine the embedding dimensions

apostolis1 committed:
Adds automated tests for the plugin. Removes the Flask dependency and instead implements a minimal HTTP server using built-in python libraries. Testing directories restructure.

apostolis1 committed:
Modifies the HTTP mockup server to use different endpoints for different responses. This means that a single HTTP mockup server can be used for testing, minimizing configurations and mariadb server restarts.

apostolis1 committed:
Escapes the input string before it is passed to curl, to ensure it is a valid JSON string

apostolis1 committed:
Adds tests communicating with the OpenAI server and improves existing tests

apostolis1 committed:
Addresses some TODOs and minor changes

apostolis1 committed:
Adds tests for bad api host value and mismatched JSON path between the API response and what is expected

apostolis1 committed:
Fixes typo

apostolis1 committed:
Converts the input to UTF-8 before sending it to OpenAI. This makes the plugin work with any input charset

apostolis1 committed:
Changes the mockup server to verify the input received and return the appropriate response from a response mapping file. Adds tests for greek charset, both for OpenAI and mockup server

apostolis1 committed:
Adds error handling for bad input encoding and corresponding test

apostolis1 committed:
Code refactoring

Exits early when the model is invalid. Changes the order of code to group relevant code pieces together

apostolis1 committed:
Adds status variables for total number of curl requests and successful number of curl requests

apostolis1 committed:
Minor change to mockup server, return a valid byte sequence in case of input not found

apostolis1 committed:
Fixes bug where a pointer to cs_openai was saved to api_response, causing a segfault

apostolis1 commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+#include "json_lib.h"
+
+static char *host, *api_key;
+ulonglong curl_requests= 0; /* Number of total curl requests */
```
Adding this comment from another commit here as well for transparency
> HTTP requests, it is called. curl is not a protocol, it is very much possible to implement the same without curl

This makes sense, will probably replace `curl` with `http`

apostolis1 replied:
Yes, I am referring to the status variable name - should have put the comment near the  `status_variables[] array`

apostolis1 commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+    {"text-embedding-3-large", 3072}
+  };
+
+  int make_openai_request() {
```
The `make_openai_request` method is a rather big method, but it does what it promises. It could be split into smaller methods if needed.

apostolis1 committed:
Adds test for status variables

The test is currently skipped because the plugin's status variables are not being reset by flush status

vuvova commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+    return 0;
+
+  c_path->p.types_used= JSON_PATH_KEY_NULL;
+  const uchar* s_p = (const uchar*) JSON_EMBEDDING_PATH.c_str();
```
why do you create a static std::string, if all you ever need is char* ?

vuvova commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+
+  c_path->p.types_used= JSON_PATH_KEY_NULL;
+  const uchar* s_p = (const uchar*) JSON_EMBEDDING_PATH.c_str();
+  if (s_p)
```
how can it be NULL?

apostolis1 replied:
It can't, this is a probably leftover code

vuvova commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+
+  *type= JSON_VALUE_NULL;
+
+  if (str)
```
this is redundant, `str` — as an argument of `val_str()` — can not be NULL

vuvova commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+    }
+    if (!str)
+    {
+      /* If str is NULL, we only care about the first found value. */
```
I don't understand that

apostolis1 replied:
Doesn't make sense indeed, probably leftover code from another version which modified `str`

vuvova commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+    This function is used to parse the vector from the JSON response
+    It should be called after the embedding is located, for the start and end positions 
+    of the json_engine to be known
+    The vector is expected to be in binary format
```
this was a bit confusing, took me a while to understand.
the vector that you parse is not in binary format, it's json array of numbers. the result of the parsing — that vector is in binary format.
so, better say "the output vector" or "the result vector" or something along those lines

vuvova commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+
+  if (str)
+  {
+    str->set_charset(js->charset());
```
not really? `str` is the resulting vector, not json. it should be always `my_charset_bin`, as far as I understand.

apostolis1 committed:
Removes CURL from required packages

apostolis1 committed:
Avoids copying the charset by using a pointer to it

apostolis1 committed:
Moves responses dict to global scope

apostolis1 committed:
Fixed copyright date

apostolis1 committed:
Removes unnecessary charset conversion

json_escape can do the conversion internally, there is no need for it to be done separately

apostolis1 committed:
Makes JSON_EMBEDDING_PATH a char* instead of std::string

apostolis1 committed:
Initialize CURL pointers to NULL

This ensures that calls to curl_easy_cleanup and curl_slist_free_all succeed

apostolis1 committed:
Changes "paths" allocation to be done in execution arena

apostolis1 committed:
Uses a String to store the post_fields

This means that the post_fields buffer can be reused for different rows, and only allocate memory when necessary. It also avoids manual memory management.

apostolis1 committed:
Changes copyright year

apostolis1 committed:
Changes the status variable names to "http" instead of "curl"

apostolis1 committed:
Implements a better waiting logic for the mockup API server

The print statements in python need to be changed to flush the output

apostolis1 commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+    // This might allow us to avoid a request if the model is not supported
+    if (model_dimensions.find(model_name) == model_dimensions.end()) {
+      my_printf_error(1, "GENERATE_EMBEDDING_OPENAI: "
+        "Model %s is not supported", ME_ERROR_LOG | ME_WARNING, model_name.c_str());
```
Should we just exit then, or should the error be logged somewhere else?

apostolis1 committed:
Cleanup of Item_func_gen_embedding::read_json

apostolis1 committed:
Minor refactoring to conform to the MariaDB coding standards

apostolis1 committed:
Removes Model is not supported error from the error log

FooBarrior commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,452 @@
+    CHARSET_INFO *cs_openai= &my_charset_utf8mb4_general_ci;
+    CURLcode ret;
+    CURL *hnd=NULL;
+    std::ostringstream read_data_stream;
```
add #include <sstream>

apostolis1 committed:
WIP Commit, this will be changed in the future and is commited only for sharing purposes

apostolis1 committed:
WIP commit, implements avoiding recomputation in alter table using the write_set.

Has some pending issues with alter statements that change the expression and type of vcols.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=4096) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=4096) before we can accept your contribution.<br/>**4** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: spetrunia<br/>:white_check_mark: drrtuy<br/>:white_check_mark: apostolis1<br/>:white_check_mark: vaintroub<br/>:x: vuvova<br/>:x: montywi<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=4096) it.</sub>

apostolis1 committed:
Changes calls to walk processors to use WALK_SUBQUERY to conform to changes from main

apostolis1 committed:
Adds check for vcol expression change in alter table

apostolis1 committed:
Casts the result of bitmap_is_set to a bool to fix 'unsafe mix of type bool and type uint'

FooBarrior commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,452 @@
+#include <mysql/plugin_function.h>
+#include <curl/curl.h>
+#include "sql_type_vector.h"
+#include "item_jsonfunc.cc"
```
Of course, the developer couldn't predict a usage outside their scope

apostolis1 commented on sql/item.cc:
```diff
@@ -965,6 +965,27 @@ bool Item_field::register_field_in_write_map(void *arg)
+{
+  TABLE *table= (TABLE *) arg;
+  bool res = false;
+  if (field->table == table || !table)
```
Ah I see.
But still, I think refactoring to `check_field_in_map` would allow checking if the field is in other bitmaps potentially.

FooBarrior replied:
Here's another point: that line can't be tested by any test case.

I'll add another point on top of it: why not to check the `field->table->write_set` overall? Do you know, would the caller want checking only in some given table's write_set, or in any table's write_set?

I'll explain. We have a multi-table update option:
UPDATE t1, t2 set t2.x = t1.y, t1.x = t2.y;

I guess in such scenarios the caller might want checking if the field is in it own table's write_set.
Though AFAIK multi-update creates a new fake table where the update happens, and then splits the results into real tables (@vuvova fix me if I'm wrong), My refactoring suggestion is based on that (thoug I can be wrong indeed!).

I'd suggest converting this check into DBUG_ASSERT(...). This way it will be useful, won't provoke extra questions, and besides we'll find out if I'm right about multi-update! :grin:

apostolis1 committed:
Moves declaration of path_exact to item_jsonfunc.h and fixes includes for plugin.cc

apostolis1 committed:
Adds a check for constant NULL values as model names and appropriate testing

apostolis1 committed:
Changes the check of table in Item_field::check_field_in_map to DBUG_ASSERT

apostolis1 commented on sql/sql_table.cc:
```diff
@@ -12698,6 +12708,54 @@ copy_data_between_tables(THD *thd, TABLE *from, TABLE *to,
+        /* 
+          If there is no change in the dependencies and the vcol expression and type (see TODO below), we can copy directly
+        */
+        if (!to->check_dependencies_in_write_set(*ptr) && def->field->vcol_info && (*ptr)->vcol_info->is_equal(def->field->vcol_info, true))
```
This check does not detect changes in the column's data type, for example changes like
`alter table t modify embedding vector(2000);`
What would be a good way to detect those?

apostolis1 committed:
Changes expensive tests to use the difference of values of the  'Generate_embedding_openai_successful_http_requests' variable

This allows them to be executed no matter the initial value of the Generate_embedding_openai_successful_http_requests status variable. It also means there is no need to reset the variable.

apostolis1 committed:
Fixes bug when assigning plugin session variables

apostolis1 committed:
Adds initial expensive test for replication setups

apostolis1 committed:
Adds dummy testing data for the plugin

apostolis1 committed:
Avoids recomputing expensive virtual stored columns in UPDATE statements

This is the initial implementation to avoid recomputing expensive virtual stored columns on UPDATE statements. This commit includes:
- Implementation of processors and methods for walking through expensive stored virtual columns and their dependencies.
- Relevant test cases using the gen_embedding plugin. Those tests need to be run with the --force-restart flag for now, due to status variables not being reset.

apostolis1 committed:
Bundles the logic of different endpoints for the mock server to the /success endpoint

apostolis1 commented:
For anyone interested, I have created [this gist](https://gist.github.com/apostolis1/8c17712a256f86969efb7d8e4cb4c591) as well with some additional information on this work.

apostolis1 committed:
Removes comment from the test

vuvova commented:
/gemini review

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+
+    decimals= 0;
+    MEM_ROOT *root= thd->mem_root;
+    json_path= (json_path_with_flags *) alloc_root(root, sizeof(json_path_with_flags));
```
Either use `thd->alloc<json_path_with_flags>(1)` or `thd_alloc(thd, sizeof(json_path_with_flags))`.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    The vector being parsed is a JSON array of numbers
+    The output vector is in binary format, will be stored in buf by this function
+    Returns 0 on success, 1 on error
+  */
```
Function comment should be moved before function itself.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    /* 
+      Make the request to OpenAI API to generate the embedding
+      Store the result in api_response
+    */
```
Function comment should be moved before function itself.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    curl_easy_cleanup(hnd);
+    hnd= NULL;
+    curl_slist_free_all(slist1);
+    slist1= NULL;
```
Resetting `hnd` and `slist1` is redundant. `cleanup` label duplicates preceding code, this can be avoided.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    return 0;
+
+  json_path->p.types_used= JSON_PATH_KEY_NULL;
+  const uchar *s_p= (const uchar *) JSON_EMBEDDING_PATH; 
```
Redundant variable.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+  json_path->p.types_used= JSON_PATH_KEY_NULL;
+  const uchar *s_p= (const uchar *) JSON_EMBEDDING_PATH; 
+  if (json_path_setup(&json_path->p,&my_charset_utf8mb4_general_ci, s_p,
+                    s_p + strlen((const char *) s_p)))
```
`strlen()` can be avoided.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/bad_response.test:
```diff
@@ -0,0 +1,20 @@
+# without examining the response body
+
+--replace_result $OPENAI_PORT OPENAI_PORT
+eval set generate_embedding_openai_host="http://127.0.0.1:$OPENAI_PORT/success";
```
Replace with `evalp` all similar occurrences.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/include/have_gen_embedding.opt:
```diff
@@ -0,0 +1 @@
+--plugin-load-add=$GEN_EMBEDDING_SO
```
Should be moved to `suite.opt`?

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/include/have_gen_embedding.inc:
```diff
@@ -0,0 +1,9 @@
+# Helper function
+let $get_expensive_calls= (select variable_value 
+   from information_schema.session_status 
+   where variable_name like 'Generate_embedding_openai_successful_http_requests');
```
We don't normally put helper functions into `have_*` files.
`like` is redundant here, use `=`.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/minimal_socket_server.py:
```diff
@@ -0,0 +1,83 @@
+    while True:
+        client_connection, client_address = server_socket.accept()
+        with client_connection:
+            request = client_connection.recv(4096).decode('utf-8')
```
IIUC, this code does not handle HTTP requests that arrive in multiple chunks, which may lead to sporadic test failures.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/minimal_socket_server.py:
```diff
@@ -0,0 +1,83 @@
+    with open(filename, 'r') as file:
+        content = file.read()
+
+    return build_response(content, status_code, content_type="application/json")
```
`get_response` looks like unused function?

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/minimal_socket_server.py:
```diff
@@ -0,0 +1,83 @@
+        200: "OK",
+        400: "Error",
+        404: "Not Found"
+    }
```
500 is missing here

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/suite.pm:
```diff
@@ -0,0 +1,64 @@
+  $skip{'t/openai_server.test'}="No OpenAI API key" unless -r $api_key_file;
+  $skip{'t/openai_greek_charset.test'}="No OpenAI API key" unless -r $api_key_file;
+  %skip;
+}
```
I'd rather make `include/openai_api_key.inc` that'd do `skip` and read some file that contains the key and then do appropriate step to set it up.

`$ENV{HOME}/openai_api_key.inc` contents is a mystery for us, we can only guess which commands it is supposed to execute.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/my.cnf:
```diff
@@ -0,0 +1,8 @@
+
+[openai]
+listen = @ENV.OPENAI_PORT
+api_response_filename = @ENV.MTR_SUITE_DIR/json_responses/api_responses.json
```
`api_response_filename` looks redundant. I'm unsure if `listen` is required too.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/greek_mockup.test:
```diff
@@ -0,0 +1,18 @@
+source include/have_gen_embedding.inc;
+--disable_query_log
+source $HOME/openai_api_key.inc;
+--enable_query_log
```
This test is failing without `openai_api_key.inc`.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/suite.pm:
```diff
@@ -0,0 +1,64 @@
+    &::mtr_add_arg($args, "$ENV{OPENAI_PORT}");
+    &::mtr_add_arg($args, $success_file_path);
+    # This can be changed to include the PORT or other info, perhaps in combination with append => 0
+    $http_server_log= "openai.log"; 
```
It create `mysql-test/openai.log` in the source tree, whereas this file has to be created somewhere in `MYSQLTEST_VARDIR`.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+static MYSQL_THDVAR_STR(host,
+  PLUGIN_VAR_OPCMDARG | PLUGIN_VAR_MEMALLOC,
+  "OpenAI API host, can be set to 'https://api.openai.com/v1/embeddings' or a custom endpoint",
+  NULL, NULL, "");
```
Given that this plugin is implemented specifically for OpenAI, shouldn't it have sane default `host` value?

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    slist1= curl_slist_append(slist1, authorization.c_str());
+    slist1= curl_slist_append(slist1, "Content-Type: application/json; charset=utf-8");
+    hnd= curl_easy_init();
+    curl_easy_setopt(hnd, CURLOPT_BUFFERSIZE, 102400L);
```
`curl_easy_setopt()` return value should be checked too, e.g. see hashicorp plugin.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    slist1= NULL;
+    slist1= curl_slist_append(slist1, authorization.c_str());
+    slist1= curl_slist_append(slist1, "Content-Type: application/json; charset=utf-8");
+    hnd= curl_easy_init();
```
`curl_easy_init()` may perform lazy global initialization by calling `curl_global_init()`, which is not thread safe. That is if multiple threads call `generate_embedding_openai()` for the first time it may cause server crash.

Another problem is that multiple plugins use of curl. Even if we move `curl_global_init()` out to plugin init routines, in theory they may call `curl_global_init()` concurrently.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/suite.pm:
```diff
@@ -0,0 +1,64 @@
+    if ($openai->{proc}) {
+      # Already started
+      return ;
+    }
```
It means single mockup API server for all mtr workers. It is probably alright for now.

One thing that I'm worried about: there can be race condition here. If multiple mtr workers hit this function simultaneously and both see empty `$openai->{proc}`.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/suite.pm:
```diff
@@ -0,0 +1,64 @@
+    &::mtr_init_args(\$args);
+    &::mtr_add_arg($args, $python_server_path);
+    &::mtr_add_arg($args, "$ENV{OPENAI_PORT}");
+    &::mtr_add_arg($args, $success_file_path);
```
I'd avoid redundant variables and pass values directly to `mtr_add_arg()`.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/my.cnf:
```diff
@@ -0,0 +1,8 @@
+api_response_filename = @ENV.MTR_SUITE_DIR/json_responses/api_responses.json
+
+[ENV]
+OPENAI_PORT = @OPT.port
```
This needs to be double checked. On my host mtr allocates 30 ports to the first worker: 19000..19029. mockup server uses port 19001. How it avoids port conflicts with e.g. replication or spider?

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/suite.pm:
```diff
@@ -0,0 +1,64 @@
+    my ($openai, $test) = @_; # My::Config::Group, My::Test
+    open(LOG, '<', $http_server_log);
+    do {
+      sleep 1 and seek LOG,0,1 until $_=<LOG>;
```
I'd do shorter `sleep(0.1)` here.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/bad_response.test:
```diff
@@ -0,0 +1,20 @@
+# responses.json includes a special entry with the key "wrong_embedding_path" for this purpose
+
+--replace_result $OPENAI_PORT OPENAI_PORT
+eval set generate_embedding_openai_host="http://127.0.0.1:$OPENAI_PORT/success";
```
`generate_embedding_openai_host` is already set to required value, so this is redundant.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_alter.test:
```diff
@@ -0,0 +1,56 @@
+insert into exp_test_new (c,d) values ("ab", "bc");
+let $expensive_calls= `$get_expensive_calls`;
+let $res= `select $expensive_calls - $expensive_calls_begin`;
+echo $res;
```
```
evalp select $expensive_calls - $expensive_calls_begin;
```

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_alter.test:
```diff
@@ -0,0 +1,56 @@
+select c, d, b, embedding is null from exp_test_new;
+
+# Should call the expensive function
+alter table exp_test_new modify c varchar(50);
```
Hmm... should it actually call the expensive function in this case? Data itself hasn't modified.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_alter.test:
```diff
@@ -0,0 +1,56 @@
+select c, d, b, embedding is null from exp_test_new;
+
+# TODO: Should this call the expensive function since the type of the vcol is changed?
+# TODO: Currently it doesn't
```
ATTN @vuvova

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_alter.test:
```diff
@@ -0,0 +1,56 @@
+echo $res;
+select c, d, b, embedding is null from exp_test_new;
+
+# TODO: Same as above, what do we want to happen here, where a dep's type changed and expression?
```
I'd say "dep's" type change as such shouldn't involve "expensive function call", unless data wasn't truncated or somehow modified otherwise.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_alter_simple.test:
```diff
@@ -0,0 +1,24 @@
+--replace_result $OPENAI_PORT OPENAI_PORT
+eval set generate_embedding_openai_host="http://127.0.0.1:$OPENAI_PORT/success";
+
+drop table if exists exp_test_new;
```
Redundant.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_alter_simple.test:
```diff
@@ -0,0 +1,24 @@
+let $expensive_calls= `$get_expensive_calls`;
+let $res= `select $expensive_calls - $expensive_calls_begin`;
+echo $res;
+select doc, x, embedding is null from exp_test_new;
```
It is essentially a copy of one of the tests from `expensive_alter.test`, right?

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_other_vcols.test:
```diff
@@ -0,0 +1,93 @@
+--replace_result $OPENAI_PORT OPENAI_PORT
+eval set generate_embedding_openai_host="http://127.0.0.1:$OPENAI_PORT/success";
+
+create table expensive_test (idx int, doc text, model text, 
```
Frankly, I’ve gotten so used to t1, t2, etc. that reading other table names hurts my eyes.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_other_vcols.test:
```diff
@@ -0,0 +1,93 @@
+echo $res;
+
+# Cleanup
+drop table expensive_test;
```
I'd move all expensive tests to a single test file.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_rpl.cnf:
```diff
@@ -0,0 +1,15 @@
+[openai]
+listen = @ENV.OPENAI_PORT
+api_response_filename = @ENV.MTR_SUITE_DIR/json_responses/api_responses.json
+wrong_json_path_filename = @ENV.MTR_SUITE_DIR/json_responses/wrong_json_path.json
```
I don't see `wrong_json_path_filename` used anywhere.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_rpl.test:
```diff
@@ -0,0 +1,68 @@
+# Test to check how expensive functions are treated in replication
```
Normally files like this are called `rpl_something`, not `something_rpl`.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/expensive_rpl.test:
```diff
@@ -0,0 +1,68 @@
+--connection slave
+source include/stop_slave.inc;
+--replace_result $OPENAI_PORT OPENAI_PORT
+eval set global generate_embedding_openai_host="http://127.0.0.1:$OPENAI_PORT/success";
```
Why slave requires this setting? ATTN @vuvova

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/greek_mockup.test:
```diff
@@ -0,0 +1,18 @@
+set @model="text-embedding-ada-002";
+--disable_query_log
+set @text="�� ��� ��� ���� ��������� ��������� ��� ������� ��� ����������� ������������� �������� ��� ����� ����� ������ � �������� ��������� ���� ���������� ��� ��������.";
+--enable_query_log
```
I'd say disabling query log in this case can be misleading.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/openai_greek_charset.test:
```diff
@@ -0,0 +1,20 @@
+
+if (!$GEN_EMBEDDING_OPENAI_PLUGIN) {
+  skip GEN_EMBEDDING_OPENAI_PLUGIN is not set;
+}
```
Why?

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/openai_greek_charset.test:
```diff
@@ -0,0 +1,20 @@
+--enable_query_log
+select vec_totext(generate_embedding_openai(@text, @model)) is null as is_null;
+
+set names default;
```
IIUC the test is essentially equivalent to `greek_mocup.test`, the only difference is `generate_embedding_openai_host` setting. There're facilities that allow to avoid such duplication.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/openai_server.test:
```diff
@@ -0,0 +1,48 @@
+
+if (!$GEN_EMBEDDING_OPENAI_PLUGIN) {
+  skip GEN_EMBEDDING_OPENAI_PLUGIN is not set;
+}
```
Looks redundant.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/openai_server.test:
```diff
@@ -0,0 +1,48 @@
+set generate_embedding_openai_api_key="wrong-api-key";
+
+call mtr.add_suppression("GENERATE_EMBEDDING_OPENAI: Bad http response code: 401");
+select vec_totext(generate_embedding_openai(@text, @model));
```
This test covers more than basic OpenAI server verification would need. I'd either simplify it or make it work with both mockup server and OpenAI.

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/reassign_variables.test:
```diff
@@ -0,0 +1,19 @@
+source include/have_gen_embedding.inc;
+
+# Verify that plugin session variables can be assigned multiple times without issues
+# This test is created because a previous plugin design had was handling
```
s/had//

svoj commented on plugin/gen_embedding/mysql-test/gen_embedding/t/status_variables.test:
```diff
@@ -0,0 +1,44 @@
+source include/have_gen_embedding.inc;
+
+--skip "Skipping for now due to plugin status variables not being reset by flush status"
```
This needs to be verified.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    {
+      return new (thd->mem_root) Item_func_gen_embedding(thd, arg1, arg2);
+    }
+  };
```
I doubt OpenAI promises stable embedding generation. It is unclear if it is safe for caching, but it must be unsafe for SBR and needs something like:
```
thd->lex->set_stmt_unsafe(LEX::BINLOG_STMT_UNSAFE_SYSTEM_FUNCTION);
thd->lex->uncacheable(UNCACHEABLE_RAND);
```

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+{
+  String tmp_js, api_response, tmp_str, post_fields;
+  json_path_with_flags *json_path;
+  std::unordered_map<std::string, int> model_dimensions= {
```
`static constexpr`?

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+#include "json_lib.h"
+
+ulonglong curl_requests= 0; /* Number of total curl requests */
+ulonglong successful_curl_requests= 0; /* Number of successful curl requests */
```
`static`

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    curl_easy_setopt(hnd, CURLOPT_FTP_SKIP_PASV_IP, 1L);
+    curl_easy_setopt(hnd, CURLOPT_TCP_KEEPALIVE, 1L);
+    curl_easy_setopt(hnd, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
+    curl_easy_setopt(hnd, CURLOPT_WRITEDATA, &read_data_stream);
```
I'd add also `CURLOPT_CONNECTTIMEOUT` / `CURLOPT_TIMEOUT`. Also think how to make `curl_easy_perform()` killable. Otherwise it is possible to exhaust MariaDB server with unkillable connections.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    {
+      null_value= true;
+      return 1;
+    }
```
Should we make `model` argument optional and use some sane default model?

svoj replied:
Should we also add `dimensions` argument?

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+
+  
+  int parse_vector(String *buf, json_engine_t je, CHARSET_INFO *cs, const uchar *start, const uchar *end);
+  String *read_json(String *str, json_value_types *type, char **out_val);
```
It looks inconsistent that `make_openai_request()` definition is kept here while `parse_vector()` and `read_json()` are moved out.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+
+String *Item_func_gen_embedding::read_json(String *str,
+                                          json_value_types *type,
+                                          char **out_val)
```
Looks like `type` and `out_val` are unused.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+
+class Item_func_gen_embedding: public Item_str_func 
+{
+  String tmp_js, api_response, tmp_str, post_fields;
```
`tmp_js` unused.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+                      strlen("{\"input\": \"") + strlen("\", \"model\": \"") +
+                      strlen("\",\"encoding_format\": \"float\"}") + 1);
+    strLen= max_chars * cs_openai->mbmaxlen;
+    post_fields.realloc_with_extra_if_needed(strLen);
```
This calculation should be reevaluated.
1. escape multiplier is much bigger for `json_escape()`, `st_append_escaped()` it claims value of 12
2. `strlen()` on constant strings should be avoided
3. `realloc_with_extra_if_needed()` looks excessive

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    post_fields.append(STRING_WITH_LEN("\",\"encoding_format\": \"float\"}"));
+    
+    // Prepare the request to OpenAI API
+    slist1= NULL;
```
Redundant.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,455 @@
+    curl_easy_setopt(hnd, CURLOPT_FTP_SKIP_PASV_IP, 1L);
+    curl_easy_setopt(hnd, CURLOPT_TCP_KEEPALIVE, 1L);
+    curl_easy_setopt(hnd, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
+    curl_easy_setopt(hnd, CURLOPT_WRITEDATA, &read_data_stream);
```
I support this idea.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+class Item_func_gen_embedding: public Item_str_func 
+{
+  String tmp_js, api_response, tmp_str, post_fields;
+  json_path_with_flags *json_path;
```
It doesn't have to be a pointer, right? Moreover it is a `read_json()` local thing, it doesn't have to be class member.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+  json_engine_t je;
+  json_path_t p;
+  const uchar *value;
+  int not_first_value= 0, count_path= 0;
```
Both variables are useless.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    post_fields.length(post_fields.length()+strLen); // We need to manually update the length
+    post_fields.append(STRING_WITH_LEN("\", \"model\": \""));
+    post_fields.append(model->c_ptr(), model->length());
+    post_fields.append(STRING_WITH_LEN("\",\"encoding_format\": \"float\"}"));
```
I'd rather use `base64` instead of `float`: 2-2.5x less data to transfer and much cheaper to parse.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+  json_get_path_start(&je, js->charset(),(const uchar *) js->ptr(),
+                      (const uchar *) js->ptr() + js->length(), &p);
+
+  while (json_get_path_next(&je, &p) == 0)
```
I'd avoid this loop unless we really want to handle multiple embeddings at once. We probably don't within the scope of this SQL function.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+    MEM_ROOT *root= thd->mem_root;
+    json_path= (json_path_with_flags *) alloc_root(root, sizeof(json_path_with_flags));
+    // We want to store max_dimensions floats, each 4 bytes 
+    fix_length_and_charset(max_dimensions * 4, &my_charset_bin);
```
Why not `max_dimensions * sizeof(float)`?

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+  
+  LEX_CSTRING func_name_cstring() const override
+  {
+    static LEX_CSTRING name= "GENERATE_EMBEDDING_OPENAI"_LEX_CSTRING;
```
Given that most modern LLMs expose REST APIs, I’d prefer a generic function such as `AI_EMBEDDING(data[, model])` that abstracts away provider-specific differences. This would avoid maintaining multiple plugins with nearly identical implementations.

svoj commented on plugin/gen_embedding/plugin.cc:
```diff
@@ -0,0 +1,449 @@
+static MYSQL_THDVAR_STR(api_key,
+  PLUGIN_VAR_OPCMDARG | PLUGIN_VAR_MEMALLOC,
+  "OpenAI API key",
+  NULL, NULL, "");
```
The fact that OpenAI key is exposed via session variable needs to be reevaluated.

1. value may shortly become visible to `SHOW PROCESSLIST`
2. value may be written to query log
3. unsure if it matters, but DBA can't set per-server key without exposing it to users
4. are there other ways for the value to become unintentionally visible?

svoj reviewed (changes requested):
This review spans only `plugin/gen_embedding` changes.

Coding style is inconsistent: spacing, lines over 80 characters, trailing spaces, variable names.

I'd move tests from `plugin/gen_embedding/mysql-test/gen_embedding/t/` `plugin/gen_embedding/mysql-test/gen_embedding/r/` to `plugin/gen_embedding/mysql-test/gen_embedding`.

I'd move the whole plugin either to `plugin/ai/generate_embedding_openai` or `plugin/func_generate_embedding_openai`.

svoj committed:
post-rebase fixes

svoj commented on sql/table.cc:
```diff
@@ -8320,7 +8320,33 @@ bool TABLE::mark_virtual_columns_for_write(bool insert_fl)
+          bitmap_updated= true;
+        }
+      }
+      else if (insert_fl)
```
Looks like this change is redundant after a4e4a56720c.

svoj commented on sql/sql_table.cc:
```diff
@@ -12703,6 +12718,54 @@ copy_data_between_tables(THD *thd, TABLE *from, TABLE *to,
+            If not stored, it cannot be copied
+            If not expensive, we decided for now to not copy
+          */ 
+          if ((*ptr)->vcol_info->is_stored() && (*ptr)->vcol_info->expr->is_expensive())
```
`is_expensive()` existed prior to this patch:
```
      This function is now used to prevent evaluation of expensive subquery
      predicates during the optimization phase. It also prevents evaluation
      of predicates that are not computable at this moment.
```

It is reused here with a different meaning and it spans more functions like `get_lock()`, `sleep()`, UDF, etc.

Do we need to re-evaluate stored virtual columns at all?

svoj reviewed (changes requested):
`sql/` comments inline.

svoj commented:
/gemini review sql/
